CARDS-2682: Introduce a new property for numeric questions: disableNegativeInput

marta- opened the pull request:
Specs: https://phenotips.atlassian.net/browse/CARDS-2682

To test:
* In the designer, create a decimal (or long) question 
* Switch on `disableNegativeInput`
* Test that the user can't enter negative numbers
* Add an answer option with a negative `value`
* Test that selecting the answer option saves the negative value in the answer

marta- committed:
CARDS-2682: Introduce a new property for numeric questions: disableNegativeInput

marta- commented:
This change would improve the user interaction with the Lab Results forms updated in https://github.com/data-team-uhn/corpath/pull/22 .

sdumitriu commented on modules/data-model/forms/api/src/main/resources/SLING-INF/nodetypes/forms.cnd:
```diff
@@ -165,6 +165,10 @@
 
+  // For number answers, disallow entering negative numbers in the inputs.
+  // This does not prevent saving negative numbers via answer options or APIs.
+  - disallowNegatives (boolean)
```
Sounds like we kind of allow negatives. But I can't think of a better name either.

marta- replied:
>Sounds like we kind of allow negatives.

I don't see what you mean.

> But I can't think of a better name either.

`disallowNegativeInput` ? `onlyNonNegativeInput`?

sdumitriu replied:
> > Sounds like we kind of allow negatives.
> 
> I don't see what you mean.

"does not prevent saving negative numbers"

> > But I can't think of a better name either.
> `disallowNegativeInput` ? `onlyNonNegativeInput`?

Too long.

marta- replied:
`disableNegativeInput` ?

marta- replied:
Addressed in 70bcc7e .

veronikaslc commented:
Tested

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/NumberQuestion.jsx:
```diff
@@ -131,7 +131,7 @@ const useSliderStyles = makeStyles()(theme => ({
 function NumberQuestion(props) {
   const { existingAnswer, errorText, classes, pageActive, disableValueInstructions, ...rest} = props;
-  const { dataType,displayMode, minAnswers, minValue, maxValue, disableMinMaxValueEnforcement, messageForValuesOutsideMinMax, isRange, unitOfMeasurement,
+  const { dataType,displayMode, minAnswers, minValue, maxValue, disallowNegatives, disableMinMaxValueEnforcement, messageForValuesOutsideMinMax, isRange, unitOfMeasurement,
```
```suggestion
  const { dataType,displayMode, minAnswers, minValue, maxValue, disableNegativeInput, disableMinMaxValueEnforcement, messageForValuesOutsideMinMax, isRange, unitOfMeasurement,
```

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/NumberQuestion.jsx:
```diff
@@ -241,7 +241,7 @@ function NumberQuestion(props) {
     min: minValue,
     max: maxValue,
-    allowNegative: (typeof minValue === "undefined" || minValue < 0 || disableMinMaxValueEnforcement),
+    allowNegative: (typeof minValue === "undefined" || minValue < 0 || disableMinMaxValueEnforcement) && !disallowNegatives,
```
```suggestion
    allowNegative: (typeof minValue === "undefined" || minValue < 0 || disableMinMaxValueEnforcement) && !disableNegativeInput,
```

marta- commented on modules/data-entry/src/main/frontend/src/questionnaireEditor/Question-hints.json:
````diff
@@ -3,6 +3,7 @@
   "expression": "Supports javascript syntax, with special placeholder variables being populated with answers to other questions in the form.\n\nExample:\n\n```return @{laps} * @{lapLength:-100}```\n\n where `laps` and `lapLength` are the names of questions in the same form, `:-` is an optional marker for a default value, and `100` is the optional default value.",
   "enableUserEntry": "If enabled, the user is not restricted to using one of the suggestions in the dropdown, and may save their own text as the free-form answer to the question.",
   "disableMinMaxValueEnforcement": "If set, the value is not _required_ to be within the specified (**Min value**, **Max value**) limits, and an out of answer value will NOT be marked as invalid. A message will be displayed under the input, warning the user that the value they entered appears abnormal, without further consequences.",
+  "disallowNegatives": "If set, the user is prevented from entering negative numbers in a numeric question input. Negative numbers can still be values of predefined answer options.",
````
```suggestion
  "disableNegativeInput": "If set, the user is prevented from entering negative numbers in a numeric question input. Negative numbers can still be values of predefined answer options.",
```

marta- commented on modules/data-entry/src/main/frontend/src/questionnaireEditor/Question.json:
```diff
@@ -133,6 +133,7 @@
           "user": {
             "minAnswers": "long",
             "maxAnswers": "long",
+            "disallowNegatives" : "boolean",
```
```suggestion
            "disableNegativeInput" : "boolean",
```

marta- commented on modules/data-entry/src/main/frontend/src/questionnaireEditor/Question.json:
```diff
@@ -197,6 +198,7 @@
           "user": {
             "minAnswers": "long",
             "maxAnswers": "long",
+            "disallowNegatives" : "boolean",
```
```suggestion
            "disableNegativeInput" : "boolean",
```

marta- commented on modules/data-entry/src/main/frontend/src/questionnaireEditor/Question.json:
```diff
@@ -263,6 +265,7 @@
           "user": {
             "minAnswers": "long",
             "maxAnswers": "long",
+            "disallowNegatives" : "boolean",
```
```suggestion
            "disableNegativeInput" : "boolean",
```

marta- commented on modules/data-model/forms/api/src/main/resources/SLING-INF/nodetypes/forms.cnd:
```diff
@@ -165,6 +165,10 @@
 
+  // For number answers, disallow entering negative numbers in the inputs.
+  // This does not prevent saving negative numbers via answer options or APIs.
+  - disallowNegatives (boolean)
```
```suggestion
  - disableNegativeInput (boolean)
```

marta- commented:
Rebased to resolve merge conflict, squashed property rename commit into the previous commit for a cleaner commit history. Please retest to make sure there are no regressions/

sdumitriu commented on modules/data-model/forms/api/src/main/resources/SLING-INF/nodetypes/forms.cnd:
```diff
@@ -165,6 +165,10 @@
   // pressing ENTER
   - enableSeparatorDetection (boolean)
 
+  // For number answers, disallow entering negative numbers in the inputs.
+  // This does not prevent saving negative numbers via answer options or APIs.
+  - disableNegativeInput (boolean)
+
```
I would put this under min-maxValue.
